### PR TITLE
Catch unhandled exception in video client scope

### PR DESCRIPTION
**Pull request by aleksandar-apostolov**
# TL;DR 
Added scope error handler to catch all `scope` exceptions and don't crash the SDK.

# Details
When a socket error would occur be it 401 or any other, it would eventually end up in one of our error flows. But in case the socket is disconnected with a valid token which in the meantime becomes invalid trying to connect the token would return `connection.error` event. This event is propagated correctly to the error flows, however the `continuation` also needs to resume and is resumed with this exception. Since there is no exception handling for this, the coroutine would crash and take the SDK down as well as the integrating app.
This was happening with very limited tokens i.e. one that expires in 1-2min.

### CODE REVIEW

File: stream-video-android-core/api/stream-video-android-core.api [hunk 1]
```diff
@@ -4460,6 +4458,10 @@ public final class io/getstream/video/android/core/sounds/Sounds {
 }
 
 public final class io/getstream/video/android/core/utils/AndroidUtilsKt {
+	public static final fun safeCall (Ljava/lang/Object;Lkotlin/jvm/functions/Function0;)Ljava/lang/Object;
```
Comment (aleksandar-apostolov) [inline stream-video-android-core/api/stream-video-android-core.api:4461]: We need to make this `internal` along with the other safe calls.

File: stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt [hunk 2]
```diff
@@ -84,3 +93,75 @@ internal fun <T> combineComparators(vararg comparators: Comparator<T>): Comparat
         comparator.thenComparing(combined)
     }
 }
+
```
Comment (aleksandar-apostolov) [inline stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt:93]: Make these functions `internal`.

File: stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/ClientState.kt [hunk 3]
```diff
@@ -16,10 +16,12 @@
 
 package io.getstream.video.android.core
 
-import android.content.Context
+import android.util.Log
```
Comment (aleksandar-apostolov) [inline stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/ClientState.kt]: Use `logger by taggedLogger` instead

File: stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/ClientState.kt [hunk 4]
```diff
@@ -80,11 +84,21 @@ class ClientState(client: StreamVideo) {
...
+    /**
+     * Returns true if there is an active or ringing call
+     */
+    fun hasActiveOrRingingCall(): Boolean = safeCall(false) {
+        val hasActiveCall = _activeCall.value != null
+        val hasRingingCall = _ringingCall.value != null
+        val activeOrRingingCall = hasActiveCall || hasRingingCall
+        logger.d { "[hasActiveOrRingingCall] active: $hasActiveCall, ringing: $hasRingingCall" }
+        activeOrRingingCall
+    }
```
Comment (skydoves) [inline stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/ClientState.kt:96]: Is there a specific reason we need to use `safeCall` here? It seems unlikely that an exception would occur in this context.

File: stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/socket/PersistentSocket.kt [hunk 5]
```diff
@@ -415,30 +419,94 @@ public open class PersistentSocket<T>(
...
+        logger.d { "[sendHealthCheck] Sending..." }
         val healthCheckRequest = HealthCheckRequest()
         socket?.send(healthCheckRequest.encodeByteString())
     }
 
+    /**
+     * Check if the socket is in a state that can connect.
+     */
+    internal fun canConnect(): Boolean = safeCall(false) {
+        val connectionState = connectionState.value
```
Comment (skydoves) [inline stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/socket/PersistentSocket.kt:441]: Same here, do we need to use `safeCall` here and the similar methods below?

File: stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt [hunk 6]
```diff
@@ -14,6 +14,12 @@
  * limitations under the License.
  */
 
+@file:OptIn(
+    ExperimentalContracts::class,
+    ExperimentalContracts::class,
+    ExperimentalContracts::class,
```
Comment (skydoves) [inline stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt]: It seems to be duplicated by mistake

File: stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt [hunk 7]
```diff
@@ -84,3 +93,75 @@ internal fun <T> combineComparators(vararg comparators: Comparator<T>): Comparat
...
+/**
+ * Safely call a function and handle exceptions.
+ *
+ * @param block the function to call.
+ */
+fun safeCall(block: () -> Unit) {
+    contract {
+        callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+    }
+    try {
```
Comment (skydoves) [inline stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt]: We can make this inline function

File: stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt [hunk 8]
```diff
@@ -84,3 +93,75 @@ internal fun <T> combineComparators(vararg comparators: Comparator<T>): Comparat
...
+ * Safely call a function and handle exceptions.
+ *
+ * @param default the default value to return in case of an exception.
+ * @param block the function to call.
+ */
+fun <T> safeCall(default: T, block: () -> T): T {
+    contract {
+        callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+    }
+    return try {
```
Comment (skydoves) [inline stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt]: We can make this inline function as well

File: stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt [hunk 9]
```diff
@@ -84,3 +93,75 @@ internal fun <T> combineComparators(vararg comparators: Comparator<T>): Comparat
...
+ */
+suspend fun safeSuspendingCall(block: suspend () -> Unit) {
+    contract {
+        callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+    }
+    try {
+        block()
+    } catch (e: Exception) {
+        // Handle or log the exception here
+        println("Exception occurred: ${e.message}")
```
Comment (skydoves) [inline stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/utils/AndroidUtils.kt]: What about using the stream logger instead? Other methods are the same.

**Review by skydoves — COMMENTED** [review]
The primary idea for fixing this issue looks good! I left some trivial comments.

File: stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/ClientState.kt (hunk 4 above)
Comment (aleksandar-apostolov) [reply stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/ClientState.kt:96]: Because we are accessing the state flows which can crash, I don't want to have a case where the flow is unavailable and it crashes, just want to return false in that case.
